### PR TITLE
[Draft PR] Relay IR visualizer.

### PR DESCRIPTION
**Pull request by chiwwang**
Hi,
When using Relay IR as a frontend, we think it would be nice to have a reasonable visualization tool.
This draft PR follows:
1. https://github.com/apache/tvm/pull/4370
2. https://tvm.apache.org/2020/07/14/bert-pytorch-tvm
3. https://discuss.tvm.apache.org/t/rfc-visualizing-relay-program-as-graph/4825

We try to define an interface Plotter to enable different plotting library.
Currently [Bokeh](https://docs.bokeh.org/en/latest/index.html) is chosen due to its interactive capability.

Example output:
https://github.com/chiwwang/tvm/blob/tmp4viz/python/tvm/contrib/relay_viz/resnet18.html

README:
https://github.com/chiwwang/tvm/blob/relayviz/python/tvm/contrib/relay_viz/README.md

We want to have your thoughts on this topic. More details are in the [RFC](https://discuss.tvm.apache.org/t/rfc-visualizing-relay-program-as-graph/4825).

Thanks:)

Thanks for contributing to TVM!   Please refer to guideline https://tvm.apache.org/docs/contribute/ for useful information and tips. After the pull request is submitted, please request code reviews from [Reviewers](https://github.com/apache/incubator-tvm/blob/master/CONTRIBUTORS.md#reviewers) by @ them in the pull request thread.

### CODE REVIEW

**Comment by echuraev** [comment]
I like this feature! It is very useful. I quickly took a look on your PR and I have one question. 
I think it will be useful to display some information on the nodes of the graph without moving the cursor on it. What do you think about it?

As an example, this is how it looks in your PR:
изображение

And the next image I took from OpenVINO documentation:
изображение

I think, it is very comfortable when you can see name of the layer, shape and some other parameters without moving the cursor on the layer. What do you think about it?

**Comment by chiwwang** [comment]
Thanks @echuraev @comaniac !
Great suggestions! I will add information on the node rectangle.
This PR will keep as draft until some of following TODOs done.
1. Display suitable information(might be configurable) on the node by default.
2. Get rid of networkx dependency.
3. Hopefully, get rid of graphviz-dev dependency.
4. Interactive legend.

File: python/tvm/contrib/relay_viz/__init__.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+            plotter.node(node_id, "TupleGetItem", "idx: {}".format(node.index))
+            plotter.edge(node_to_id[node.tuple_value], node_id)
+        elif isinstance(node, tvm.ir.Op):
+            pass
+        elif isinstance(node, relay.Let):
+            plotter.node(node_id, "Let", "")
+            plotter.edge(node_to_id[node.value], node_id)
+            plotter.edge(node_id, node_to_id[node.var])
+        else:
+            unknown_info = "Unknown node: {}".format(type(node))
```
Comment (leandron) [inline python/tvm/contrib/relay_viz/__init__.py:101]: One issue I found when partitioning `mobilenet_v1_1.0_224_quant.tflite` with ethos-n before building it, I got:
```
Unknown node: <class 'tvm.ir.expr.GlobalVar'>
```

**Review by leandron — COMMENTED** [review]
I applied your PR locally. The output looks awesome.

A few comments and suggestions below.

File: python/tvm/contrib/relay_viz/README.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+apt-get install graphviz graphviz-dev
+# pygraphviz
+pip install pygraphviz==1.6
+
+# networkx
+pip install networkx==2.5.1
+
+# bokeh
+pip install bokeh==2.3.1
+```
```
[the remark this answers is not in the document]
Comment (chiwwang) [reply python/tvm/contrib/relay_viz/README.md:50]: Thanks @leandron !
I will check that .py after I stripped out at least networkx dependency (though I am thinking it is possible to get rid of all graphviz-dev, pygraphviz and networkx dependency, still need to figure out DOT language.)

File: python/tvm/contrib/relay_viz/__init__.py (hunk 1 above)
Comment (chiwwang) [reply python/tvm/contrib/relay_viz/__init__.py:101]: Ah, Thanks! will handle global var.

**Comment by elvin-n** [comment]
@chiwwang that's great feature, thank you for adding it!
some comments

- Is my impression correct that visual representation can mix the order of the inputs into the node? I.e. I look into quantize op and it accepts tensor to quantize by 0th port, but sometimes this tensor is visualized between scale and zero point. Can we add the ordering anyhow?
- sometimes edges are drown coming to/ going out from sides of the box and it is unclear if it is input edges or output. Can we add arrows on the edges (more preferable) or make input data coming only from top and output data going only from bottom?

**Comment by chiwwang** [comment]
@elvin-n Thanks for suggestions!
1. The position and edge are provided by graphviz and after some googling.. yes it will mix input order :(
https://stackoverflow.com/questions/44274518/how-can-i-control-within-level-node-order-in-graphvizs-dot
I will add this to TODO and investigate how to handle this as above StackOverflow discussed (i.e node attribute "ordering")

For arrow-head, I am thinking to add a triangle at the end of the edge. This should hopefully not a big problem. Will try to add this :)

**Comment by chiwwang** [comment]
A quick status update. Following functionalities are done:
1. Show node-information with the zoom-in/out level of the plot.
2. Interactive legend, clicking the legend item to highlight a node-type.
Because Relay don't have name for each nodes, this functionality can somehow serve an alternative of searching. :)
3. One `GlobalVar`, one graph. A .html file can contains multiple graphs, arranged as a column.

Let's wait for conclusion about interfaces in #8668. I will turn this Draft PR to a PR for reviewing then.
Thanks :)

Example:
[Resnet18](https://github.com/chiwwang/tvm/blob/54043c815513268d653e137c9cb23c04f783a591/test_relayviz.html)
[Multiple GlobalVar](https://github.com/chiwwang/tvm/blob/54043c815513268d653e137c9cb23c04f783a591/test_relayviz_multiple_globalvar.html)

**Comment by chiwwang** [comment]
#8668

**Pull request closed without merging**